[RFC] Mcuboot XIP Support for Zephyr

de-nordic opened the pull request:
Provided script will automatically build alternative slot application.

Invocation:
python3 zephyr/scripts/direct_xip.py <original_build>

TODO:
- [ ] add destination path selection instead of using harcoded one
- [x] error/exception handling

de-nordic commented:
Mcuboot says it works, smp_svr sample executed from slot 2
```
uart:~$ *** Booting Zephyr OS build zephyr-v2.4.0-1312-g2f4b66967d31  ***
[00:00:00.256,378] <inf> mcuboot: Starting bootloader
[00:00:00.262,329] <inf> mcuboot: Primary slot: Image not found
[00:00:00.268,951] <inf> mcuboot: Secondary slot: version=0.0.0+0
[00:00:00.416,137] <inf> mcuboot: Booting image from the secondary slot
[00:00:00.423,492] <inf> mcuboot: Bootloader chainload address offset: 0x73000
[00:00:00.431,488] <inf> mcuboot: Jumping to the first image slot
*** Booting Zephyr OS build zephyr-v2.4.0-1312-g2f4b66967d31  ***
[00:00:00.005,859] <inf> smp_sample: build time: Nov 12 2020 15:10:12
```

Dump from mcumgr:
```
$$ mcumgr_serial image list
Images:
 image=0 slot=1
    version: 0.0.0
    bootable: true
    flags:
    hash: b29b36984a7e4d2a83d4cfe34d53ff96d16bb75b85cdbf24f02ff8d73ec32807
Split status: N/A (0)
```

utzig commented on Kconfig.zephyr:
```diff
@@ -80,20 +80,46 @@ config HAS_FLASH_LOAD_OFFSET
-	  below.
+	  When enabled, the choice of FLASH_LOAD_OFFSET/FLASH_LOAD_SIZE will
+	  be restricted to fixed partitions defined in device tree.
+	  By default the chisen zephyr,code-partition will be used.
```
typo: `chisen` -> `chosen`

utzig commented on Kconfig.zephyr:
```diff
@@ -80,20 +80,46 @@ config HAS_FLASH_LOAD_OFFSET
+	prompt "Application partition"
+	default APP_IN_Z_CODE_PARTITION
+	help
+	  Menu allows to select parition the application will be build for.
```
typos: `parition` -> `partition`, `build` -> `built`

utzig commented on Kconfig.zephyr:
```diff
@@ -80,20 +80,46 @@ config HAS_FLASH_LOAD_OFFSET
+	bool "zephyr,code-partition \($(dt_chosen_label,$(DT_CHOSEN_Z_CODE_PARTITION))\)"
+
+config APP_IN_IMAGE_0_PARTITION
+	bool "slot0_parition"
```
typo: `slot0_parition` -> `slot0_partition`

utzig commented on Kconfig.zephyr:
```diff
@@ -450,6 +478,14 @@ config MCUBOOT_GENERATE_CONFIRMED_IMAGE
 	  and CONFIG_BUILD_OUTPUT_HEX.
 
+config MCUBOOT_TARGET_IMAGE
+	int"Selects image slot for which application will be build"
```
add a space after the keyword, typo: `build` -> `built`

de-nordic replied:
Option will be remove; this is leftover from experiments.

Laczen commented:
@de-nordic, I don't think it is required to do this change, correct usage of chosen partition allows to generate the correct image. The chosen partition info also has the required data to be passed to imgtool and to dfu-util/nrfjprog/...
To allow support for a bootloader (not only mcuboot) there are only two required pieces of info for a application image: where it will be placed in flash (partition offset) and where the image starts (unencrypted, uncompressed, ...): this is the partition offset + a image header size.

Easier support for uploading using dfu-util could be achieved by adding the location where the binary should be placed in flash to the header info, in that case the dfu "server" could  read in the header and decide where to place it or to reject it. This is much more flexible that selecting alt-0 or alt-1 (BTW, selecting alt-0 to "upload" from flash should be disallowed). When using a "hex"-loader this is maybe less required, but still usefull for the bootloader to distinguish between execute or move.

de-nordic commented:
> 
> Easier support for "downloading" using dfu-util could be achieved by adding the location where the binary should be placed in flash to the header info in that case the dfu "server" could read in the header and decide where to place it or to reject it.

Yes, but you are talking abut hi-end solution, I am still trying to build things for two slots. You may add that comment to the original issue reported here: https://github.com/zephyrproject-rtos/zephyr/issues/27673.

But yes, it would be nice to add header that would identify the slot of binary and allow dfu client to reject it citing the problem, or even pack two binaries together (one for each slot) and allow dft to negotiate which one will go in.

I will try with  `zephyr.dfu-code-partition` that is supposed to be set to the slot the application will be uploaded to.
The combination of `zephyr,code-partition` and `zephyr, dfu-code-partition`, or lack of the former one, would decide whether application is built for dual slot bootable device, XIP dual boot device or single slot device.

Laczen commented:
@de-nordic, maybe I am missing something, I was thinking you are trying to create a image to be executed from slot1. This could be done by setting the chosen partition to slot1 (or image1) nothing more. But then again, maybe you are trying to achieve something else.

nandojve commented:
Hi folks, I noted this improvement and have couple questions:

1) How MCUboot will handle 3 or 4 images, or even more?

I mean, there are work on Zephyr to enable NS and SE images (non secure and secure images). These are two independent partitions + at least another one as temporary storage (scenario 1). So, how MCUboot + Zephyr will handle this ? I imagine XIP is essential for lots of platforms and there is a special case, i.MX-RT that only have external memory. Having two images on a single bin (scenario 2) is an alternative but can be impractical with LWPAN. My firsts tests with BLE IPSP (PC <-> nRF) needs around 30min to download and update 1 image with minimal resources to run the example without mesh and only 2 nodes.

2) Since there will be needed work to change the current dual image to a new thing, would be interesting have a infrastructure that allows multi-image support, or something that address this current/future scenario?

Sorry if this is not the right place to add both questions but it seemed opportune to raise them.

CC: @otavio

utzig commented:
> 1. How MCUboot will handle 3 or 4 images, or even more?

The `mcumgr` CLI app already has a flag to choose the image pair where to upload, `image upload -n <image-number>`, which can be used to choose between S/NS, but this currently uploads in a fixed slot. I don't remember if the mcumgr library supports this, but MCUboot surely does. With the XIP support it will probably also get an extra flag for the slot number. Then a user will have control of what image pair and slot a particular image should be uploaded to.

nvlsianpu commented on subsys/usb/class/usb_dfu.c:
```diff
@@ -344,7 +353,7 @@ static void dfu_reset_counters(void)
 	dfu_data.bytes_sent = 0U;
 	dfu_data.block_nr = 0U;
-	if (flash_img_init(&dfu_data.ctx)) {
+	if (flash_img_init_id(&dfu_data.ctx, UPDATEABLE_FLASH_IMG_ID)) {
```
why this is not applied in `dfu_data declaration`?

nvlsianpu reviewed (commented):
Did you consider to keep working usb dfu class when it is part of mcuboot in single application-slot-mode?

de-nordic commented:
@Laczen I have incorporated your suggestion regarding adding information on image destination in header have delivered prototype solution that allows to sign image for mcumgr, with imgtool, with desired address set. This way image may be rejected if it is supposed to be uploaded at specified ROM address but selected slot will have different address.

de-nordic commented:
> 
> 
> > 1. How MCUboot will handle 3 or 4 images, or even more?
> 
> The `mcumgr` CLI app already has a flag to choose the image pair where to upload, `image upload -n <image-number>`, which can be used to choose between S/NS, but this currently uploads in a fixed slot. I don't remember if the mcumgr library supports this, but MCUboot surely does. With the XIP support it will probably also get an extra flag for the slot number. Then a user will have control of what image pair and slot a particular image should be uploaded to.

The zephyr backend, of mcumgr, will attempt to select one of two images, which basically would be the non-active one.

de-nordic committed:
scripts: direct_xip.py alternative Direct-XIP application generator

The script is intended to be run on already, successfully, compleated
build to generate build for alternative slot.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

tejlmand commented:
@de-nordic please take a look here: #40555 I believe the approach in that RFC will be very suitable for this kind of work you need to do here.

de-nordic commented:
> @de-nordic please take a look here: #40555 I believe the approach in that RFC will be very suitable for this kind of work you need to do here.

Thanks, I will take a look.
